parachain/pallets/control/src/lib.rs

```diff
@@ -131,31 +156,156 @@ pub mod pallet {
 		#[pallet::call_index(1)]
 		#[pallet::weight(T::WeightInfo::create_agent())]
 		pub fn create_agent(origin: OriginFor<T>) -> DispatchResult {
-			let mut location: MultiLocation = T::CreateAgentOrigin::ensure_origin(origin)?;
+			let origin_location: MultiLocation = T::ControlOrigin::ensure_origin(origin)?;
 
-			// Normalize all locations relative to the relay chain unless its the relay itself.
-			let relay_location = T::RelayLocation::get();
-			if location != relay_location {
-				location
-					.reanchor(&relay_location, T::UniversalLocation::get())
-					.or(Err(Error::<T>::LocationConversionFailed))?;
-			}
+			let (agent_id, para_id, location) = Self::convert_location(origin_location)?;
 
-			// Hash the location to produce an agent id
-			let agent_id = T::AgentHashedDescription::convert_location(&location)
-				.ok_or(Error::<T>::LocationConversionFailed)?;
+			log::debug!(
+				target: LOG_TARGET,
+				"💫 Create Agent request with agent_id {:?}, origin_location at {:?}, location at {:?}",
+				agent_id,
+				origin_location,
+				location
+			);
 
 			// Record the agent id or fail if it has already been created
-			if Agents::<T>::get(agent_id).is_some() {
-				return Err(Error::<T>::AgentAlreadyCreated.into())
-			}
-			Agents::<T>::insert(agent_id, ());
+			ensure!(!Agents::<T>::contains_key(agent_id), Error::<T>::AgentAlreadyCreated);
+
+			Agents::<T>::insert(
+				agent_id,
+				AgentInfo { origin: VersionedMultiLocation::from(location), para_id },
+			);
+
+			let message = Message {
+				origin: T::OwnParaId::get(),
+				command: Command::CreateAgent { agent_id, para_id },
+			};
+			Self::submit_outbound(message.clone())?;
+
+			log::debug!(
+				target: LOG_TARGET,
+				"💫 Create Agent request processed with outbound message {:?}",
+				message
+			);
+
+			Self::deposit_event(Event::<T>::CreateAgent {
+				location: Box::new(location),
+				para_id,
+				agent_id,
+			});
+			Ok(())
+		}
+
+		/// Sends a message to the Gateway contract to create a new Agent representing `origin`
+		///
+		/// - `origin`: Must be `MultiLocation`
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::create_channel())]
+		pub fn create_channel(origin: OriginFor<T>) -> DispatchResult {
+			let location: MultiLocation = T::ControlOrigin::ensure_origin(origin)?;
```

Comment:
Must also verify that the origin is a toplevel parachain, I.e "/Parachain(X)", and not some nested account within a parachain.

Ie only Parachains can create channels, not ink! smart contracts, etc.

Reply:
That depends on how source chain implemente `SendXcmOrigin`. e.g. `xcm_pallet` in `asset_hub_westmint` we use for launching xcm transact tests in this PR use `LocalOriginToLocation` in https://github.com/Snowfork/cumulus/blob/5f9f6cd8eb6d779d8b713dd7f931a9521e93b132/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L457, so origin becomes `(1,X2(Parachain(x),AccountId(y))) when it reachs bridge_hub.

We may implement some a customize xcm call or a cutomize runtime which encode to /Parachain(x) but I would assume it's not required for this PR and I've some other thought regard this in https://github.com/Snowfork/snowbridge/blob/dc9c4ff0cbd7004cfdc74f3964b2a32055aaad4f/parachain/pallets/control/src/lib.rs#L325-L330

Reply:
I see the problem. Thanks Ron.

Anyway, I don't think its a huge  issue since create_channel is idempotent and is not destructive.

So the current code looks good.

Reply:
That technique works great for smoketests that test Polkadot->Ethereum messaging. However it doesn't help us in the opposite direction where for example we need to verify that assets were correctly deposited into the final destination parachain.

Reply:
https://github.com/Snowfork/snowbridge/pull/917#discussion_r1297081685

Reply:
> we really must enforce that the origin is /Parachain(X) and not /Parachain(X)/AccountId32(Y) or anything else

Actually this is already enforced in `create_channel`  and we check existence of agent when `create_channel` so implicitly applies to `create_agent`.  Reason I did not enforce that in `create_agent` is  that will break most of the test cases in Control pallet with initial intention to allow diversity of origins.

That's also the reason why I asked in https://linear.app/snowfork/issue/SNO-615#comment-445ddfb8. Aidan gave some comment about the scope for longer term and I agree with you to explictily enforce it for now to make less confusion.

Reply:
About the helper pallet, I would assume it's just some improvement and would suggest to do it in another PR.

Btw: there is already some low level modules referenced in https://github.com/open-web3-stack/open-runtime-module-library/blob/master/xcm/src/lib.rs

Reply:
Some revamp in https://github.com/Snowfork/snowbridge/pull/917/commits/eb484f72e178b96f0e078e2582f0a539bb9bfa82